### DISCUSSION TITLE
Drop Python 2 support

### DISCUSSION THREAD

**Discussion by adamjstewart** [spack/spack#31824] [General]
For anyone new to Spack development, you may be surprised to learn that Spack still supports Python 2 (both for running Spack, which is written in Python, and for installing Python libraries). However, Python 2 has reached [EOL](https://endoflife.date/python) and is no longer supported by the PSF. This introduces both risks (Spack will happily install a version of Python that may contain known security vulnerabilities) and significant maintainer burden. I would like to once again open the can of worms that is dropping support for Python 2 (and EOL Python versions in general).

I would like to split this discussion into two directions:

1. Versions of Python that can be used to run Spack
2. Versions of Python that Spack can install

These do not necessarily need to happen in tandem. When expressing opinions in favor of or against changes in Python version support, please make it clear which you mean.

### 1. Versions of Python that can be used to run Spack

At the moment, Spack supports Python 2.7 and 3.5–3.10. We only recently dropped Python 2.6 support in Spack 0.18 (#27256). I think we should accelerate this timeline and require at least Python 3.6+.

Many Spack developers (myself, @alalazo, @trws) have expressed interest in using Python language features introduced in modern Python. These include things like type hints (introduced in Python 3.5) and f-strings (introduced in Python 3.6). Most of our vendored dependencies can no longer be updated because they dropped Python 2 support years ago.

Why do we still support Python 2.7 for running Spack? Many HPC centers run clusters that use RHEL 7 which comes with Python 2.7 by default. Unfortunately, RHEL 7 is [still supported until 2026](https://access.redhat.com/product-life-cycles/). I don't think it's realistic for Spack to also continue to support Python 2.7 for that long.

What about users stuck on RHEL 7? There are two easy solutions: `yum install python3` or use Spack 0.18. Users can even use Spack 0.18 to install Python 3, then upgrade to newer versions of Spack. I personally don't believe that this is a significant hurdle to Spack usage, as Spack already has a number of [system prerequisites](https://spack.readthedocs.io/en/latest/getting_started.html#system-prerequisites) in addition to Python.

### 2. Versions of Python that Spack can install

At the moment, Spack supports installing Python 2.7 and 3.1–3.10. I think we should follow PSF and only support Python 3.7+.

While working on #27798 and #31823, I've found it quite difficult to support both Python 2 as well as modern Python. I already tried deprecating EOL Python versions in #28003 but received pushback from a user (singular) who still relies on Python 2-only libraries. 

There will always be a tradeoff between maintainability and reproducibility. The Python ecosystem is evolving so rapidly that it is becoming difficult to keep up while still also supporting Python 2.7 (released 12 years ago). I understand concerns about reproducibility, but there is an easy solution for this: use Spack 0.18. If users want to be able to install ancient versions of software, they can use ancient versions of Spack (okay, not yet ancient, but will be a decade from now when people are still requesting Python 2 support). Unless we allow deprecation and removal of ancient versions of Python, one of 2 things will happen: Spack will start to fall behind other package managers in terms of Python support, or I will go crazy trying to support everything.

In #31823, I introduce support for the `--config-settings` flag for pip. However, support for this flag was added in pip 22.1, which only supports Python 3.7+. This means that any package we add `config_settings` to will only support Python 3.7+. For `py-scipy`, we will only add this flag to scipy 1.9+, which already only supports Python 3.7+, so there's no issue there. But for other packages which specify a `build-backend` in `pyproject.toml`, this raises the question of which Python versions we are willing to support.

This is not the only feature reliant on removing support for EOL Python versions. We currently use pip to install all Python libraries. However, pip is an entire package manager with support for many things we don't need or use in Spack. Instead of our current "package manager (pip) inside of a package manager (spack)" approach, I would love to use [build](https://github.com/pypa/build) and [installer](https://github.com/pypa/installer) instead. However, these libraries also require Python 3.7+.

### Comparison with other package managers

It's always good to look outside of our own bubble and see what other package managers are doing:

* pip: 3.7+
* conda: 3.7+
* brew: 3.7+

I dream of the day that Spack can join this list of popular package managers brave enough to abandon Python 2 support.

Relevant parties: @varioustoxins @scheibelp @tgamblin @skosukhin @pradyunsg

**Comment by pradyunsg** [comment]
My 2 cents are: do it.

Running a yum command to get a newer version of Python is not that painful. And, as long as you have a mechanism for users to use the older version, they can. Unless there are users who are supporting the development burdens of EoL Python development with huge sums of money/resources dedicated to the project that justify the pains of compatibility with legacy software. IIUC, the only folks still on RHEL 7 are institutional users who have a support contract with RedHat, either directly or through a intermediary.

I’ve said this in various places, but OSS maintainers (including packaging tooling) shouldn’t feel obliged to maintain support beyond EoL if they’re not getting paid sufficiently for supporting it. And, most aren’t. Even for those that are, it still rarely makes sense to try to match RHEL’s support cycle — RedHat has both the scale and scope to employ a small full-time team for this (it’s small teams, even for foundational pieces like Python or gcc!) and, even then, they only have to support one Python version or two per RHEL version which they directly control and patch.

And… a huge chunk of the Python ecosystem moved in sync with CPython on dropping Python 3.6 support, when it hit EoL within CPython’s workflow. I won’t be surprised if the same happens for 3.7.

**Reply by pradyunsg** [reply]
See https://github.com/psf/black/issues/2251#issuecomment-885829194 for one of the other instances of me communicating this.

**Reply by tgamblin** [reply]
Maybe I should clarify something here, because this view might reflect industry vs. hobby open source, but it's not really reflecting the world of HPC accurately.  Note: I am **not** saying that I don't think we should eventually drop Python 2.7 support, but I've seen @pradyunsg write this same argument elsewhere (e.g., when Black dropped 2.7 support) and it's not a match for the world of research computing.

### Some clarifications on Pradyun's $.02

> Running a yum command to get a newer version of Python is not that painful.

If you have root permission on your machine, this is true.  Most HPC users do not have that.  **Do not assume that users of your software have root or that they can use system package managers**.  This is one of the reasons Spack even exists -- the users of HPC systems *cannot* use APT or similar to install software.

> Unless there are users who are supporting the development burdens of EoL Python development with huge sums of money/resources dedicated to the project that justify the pains of compatibility with legacy software.

It's not just people with huge sums of money who are supporting legacy software.  @varioustoxins in his prior emails points out that there are huge volumes of research code still written in Python 2.7, and those teams don't have funding to retrofit their code to Python 3.  That's one of the arguments for continuing to support the installation of Python 2.7, and I'm sympathetic to it because Spack is intended to support *research code*.  HPC *centers* may have tons of money for hardware (most don't) but you may be surprised that their users do *not* likewise have tons of money for software.  This project exists to help at least some of those teams.

> the only folks still on RHEL 7 are institutional users who have a support contract with RedHat, either directly or through a intermediary.

Similarly, no.  Many HPC users are beholden to what is provisioned for them at their center.  They don't get a lot of choice -- they run on the hardware their university or lab has, and the reason the center may not have upgraded is likely due to things like vendor support for kernel drivers for fast networks, or inability to upgrade to the latest generation of such hardware.  They may not even have software developers or computer scientists on their teams.  

Many sites also run CentOS, Rocky, etc. -- the free re-spins of RHEL.  You can see that in the OS chart here from our 2021 user survey:

![two_multi_bars_os](https://user-images.githubusercontent.com/299842/182050220-504c67b3-fad7-48e3-9d01-515367c31a54.svg)

Within DOE/NNSA (which keeps this project going), over 60% of users were still running RHEL7 at the beginning of 2022, and something like 20% were using CentOS 7.  That is very significant.  In the overall user community, it's much less pronounced, but 33% of users were targeting RHEL7 or CentOS7.  There is a unfortunately lot of inertia there in terms of what HPC centers.run.

Within NNSA, that is going to change real soon.  The thing perpetuating RHEL7 here is that our current generation of [CTS-1 clusters](https://www.energy.gov/nnsa/articles/nnsa-announces-procurement-penguin-computing-clusters-support) use it.  Those are on the way out with [CTS-2](https://www.energy.gov/nnsa/articles/nnsa-taps-dell-technologies-provide-expanded-computing-resources-stockpile) rolling out within months.  No new systems are going to use RHEL7, but CTS-1 systems have lasted a *long* time (for better and for worse).  I expect a very different look for our distribution of OS's by next year.

> OSS maintainers (including packaging tooling) shouldn’t feel obliged to maintain support beyond EoL if they’re not getting paid sufficiently for supporting it.

Spack is funded by the ASC program (which is quite large) and there are at least 6 or so of us who are getting paid to help support this stuff.  Much more if you count the support teams at labs who use Spack.  @adamjstewart is not one of those paid maintainers, and he does the lion's share of the work to support the Python side of things, so I have a lot of sympathy for these arguments (particularly his).  But Spack is far from an unfunded project.

### But we should do it anyway

All that said, based on real data, I think **it still makes a lot of sense to deprecate support for running under Python 2.7 with `0.19`, and to remove it in `0.20`** (or 1.0 or whatever we call it).  Nearly all sites now allow you to `module load python/3` and get a Python3 implementation, even if you are an unprivileged user.  The really interesting use case for using 2.7 is bare CentOS containers, but the major distros have already moved to Python3, and Spack works under RHEL8 `platform-python` (we actually test that).  More convincing is this chart from the 2021 survey:

![two_pies_how_bad_only_py3](https://user-images.githubusercontent.com/299842/182050229-d36b40c8-c104-4dfa-9a82-7c87519798b6.svg)

Only 1.09% of users are totally not cool with this.  23.5% will be ok with it b/c admins provisioning systems will do the `yum install` thing (or get their center to), and users will load a module. There aren't centers that don't provide Python 3 one way or another.  Within NNSA, **nobody** finds the switch unacceptable, though more seem to think that loading a module is a pain.  That said, it's not **that** big of a pain, and yes we should do it.

### What about supporting 2.7 in Spack?

Dropping support for 2.7 in Spack is a bit more subtle to me, because I know there are teams that want to build things in 2.7.  I think continuing to support that in 0.18 or 0.19 is fine, *provided* we get a little better at reproducibility.  I think we should start baking in the hash of the Spack repo to our environments, and enable two-step reproducible builds where we check out the spack version that built the env and build it with that version.  That would solve two issues: much more exact reproducibility and supporting old versions of things.  Basically the support model would be: you can reproduce thing with the spack version you used to build them in the first place.

Thoughts?

**Reply by adamjstewart** [reply]
@tgamblin if we add the feature you mention that bakes the Spack git commit into the environment lock file, would you be comfortable deprecating Python 2.7, 3.5–3.6 in 0.19 and dropping them in 0.20? Our release timeline is already quite slow, but I imagine the plan is to release 0.19 in November and then immediately drop it in develop after that.

I sympathize with users of scientific software whose developers don't have the time/money to port to Python 3. Honestly, I would do it for them myself if I cared less about graduating someday. In my experience, using `2to3` gets you 90% of the way there with most libraries, although the last 10% can be quite painful. This is also only for pure Python libs, I have less experience with Cython.

**Reply by tgamblin** [reply]
I am all for deprecating 3.5. Why 3.6 though? What does 3.7 get you? My motivation for keeping 3.6 is that platform-python is 3.6, so spack can work out of the box on a RHEL8 UBI container.  Here's the Python version data I have
![two_multi_bars_python_version](https://user-images.githubusercontent.com/299842/182050266-698d0c8e-a4eb-4878-88e7-53ffe0d2a040.svg)
.

**Reply by adamjstewart** [reply]
Then how about:

1. Versions of Python that can be used to run Spack: 3.6+
2. Versions of Python that Spack can install: 3.7+

1) allows us to use modern language features in Spack. 2) allows us to support new build system features like `pip install --config-settings` or build/installer when installing Python packages. Spack will no longer install EOL versions of Python that may contain security vulnerabilities.

**Reply by vsoch** [reply]
+1, I am in support of pushing for better practices to use recent versions of software (Python being a big deal). If operating systems / other managers are entirely deprecating versions, then we should too. And yes, some ornery developer is going to be angry "OHNO I have to update the things!" but they really should be updating the things. It's reasonable to have some backwards compatibility to give people time to do that, but we don't have to support old versions forever that frankly people should not be using. We are encouraging bad practices by not doing this. So +1, update spack to something modern, and tar and feather anyone that protests.

**Reply by trws** [reply]
Yes. Even if it means I need to get us a statically compiled python 3 that can run spack so we can download it on bare containers, I'm all for this. If we do that, I'd actually push for 3.7 because that gets us delayed evaluation of types, which will help with type annotations in some of the complex cases we need, dataclasses, and perhaps most importantly, the ability to set __getattr__ on modules so we can control module scope definitions in packages while transitioning away from poking. It also has far better support for generic types and doing things with declared types in dynamic code. Perhaps not critical, but some things that would legitimately help us move toward where I think we want to be.

**Reply by tgamblin** [reply]
I agree with most of this but will just reiterate my point above:

> If operating systems / other managers are entirely deprecating versions, then we should too.

Spack is not `homebrew`, `conda`, or `pip`; it's got a different community, a different reason for existing, and a different funding model from all of those.  There's obviously some overlap, and we have to keep Spack maintainable and balance the needs of maintainers and users.  If we alienate our users because we do what "everyone else" is doing when it *doesn't* make sense, we aren't helping anyone.  In this case, the most used OS is *not* deprecating one EOL python version, but other package managers are.

In this case I think it makes a lot of sense to keep moving our Python version forward, and I hope Python's pace will continue to pull the community forward.

The degree to which this community uses RHEL makes me think that Spack should *at least* run on Python 3.6, as RHEL8 is going to be around for a while.  If that gets burdensome, then I actually like @trws's suggestion better -- bring our own Python so that we don't have to worry about this stuff.  That actually allows us to run in *more* places, and it strikes a really nice balance between maintainers and users.  It also means we need to do more bootstrapping work for Python, but I think that's a good thing.  It would give us way more choice in the long run.

So yeah, I like the run on 3.6 and install 3.7 thing, for now, provided we can reproduce older envs by bootstrapping:

1. The Spack version
2. The package repos used to build the env

This is something people are integrating into wrappers around Spack anyway -- for the reproducibility.  It also makes it easier to drop things, and instead of supporting everything all the time in core, we can let people have external package repos where they support older things if they want to.

**Reply by adamjstewart** [reply]
Other ideas that no one has really talked about in this thread yet:

1. Once Spack core and packages are split into separate repos, it would be feasible for me to maintain a Python 3-only repo and someone else could volunteer to maintain a Python 2-only repo.
2. Once @alalazo's multi-build system PR is finished, that may improve some of the issues. If we can dynamically use build/installer for Python 3.7+ and pip for Python 3.6 and older, that may ease some of the support issues.

**Reply by vsoch** [reply]
> Once Spack core and packages are split into separate repos, it would be feasible for me to maintain a Python 3-only repo and someone else could volunteer to maintain a Python 2-only repo.

Holy cow is that on the horizon? I've been going on about that forever... and having each package tested and provide its own binaries.

**Reply by adamjstewart** [reply]
It's been on the horizon for at least 5 years, will probably continue to be on the horizon for several more years. I think the plan was to reconsider it once Spack 1.0 is released.

**Reply by trws** [reply]
FWIW, the bring your own python thing is what homebrew does with Ruby in many cases already.

Also, thought on the packages split thing... I ran into this the other day and started thinking how convenient it might be for things like spack, or raja-framework: https://github.com/josh-project/josh

It's a proxy that provides a standard git endpoint that lets you subset bits of a monorepo as though they were completely separate repositories, completely transparent to git clients in both directions.  Keep everything in one repo, but say just clone spack core as a whole, and packages as a shallow depth 1, for example.  Maybe not worth it, but got the gears spinning for sure.

**Reply by vsoch** [reply]
Not a new idea for me! https://syspack.github.io/pakages/

**Reply by alalazo** [reply]
Talking about 1. (which version we should use to run Spack), here's a PR that makes me want to have support for Python 3.6+:
- https://github.com/spack/spack/pull/31393

Why do we need to vendor old backports to `pathlib` (a standard library package) when we can just:
```python
import pathlib
```
?

**Reply by alalazo** [reply]
Another kind of issues that will disappear when we'll move to Python 3 only: CI failures on Python 2.7 due to UTF-8 characters that are not decoded correctly. See https://github.com/spack/spack/pull/32608, which caused one job in CI to fail consistently.

**Reply by alalazo** [reply]
Another commit that caused issues on Python 2.7 due to UTF-8 characters https://github.com/spack/spack/commit/71434d8b9d05e5d89a57ba992ded2d22ef6e6765

**Comment by adamjstewart** [comment]
Update for 1: Spack no longer supports running Spack with Python 3.5: https://github.com/spack/spack/pull/31908

**Comment by adamjstewart** [comment]
Update for 1 and 2: Python 2 support will be officially deprecated once #32615 is merged.

**Comment by scheibelp** [comment]
Regarding (2)

I think installing Python 2.7 with Spack 0.18/0.19 (according to https://github.com/spack/spack/pull/32615#issue-1370300748) is a good solution. I want to consider whether there is an easy path for users that need to install a mixture of packages needing Python 2.7 and newer packages that may only be available in a later Spack version (i.e. as new packages are added to Spack > 0.19). 

To do that, a user could install Python 2.7 with Spack 0.19 and then create a manifest .json file (or externals in `packages.yaml`) for a separate Spack >= 0.20 installation. This assumes that all `py-*` packages are installed with Spack 0.19; that generally seems reasonable other than for `py-*` packages that need e.g. `blas` (even then, I assume there will eventually be API deviations that constrain the `blas` version used).

Another option is to improve support for switching Spack versions within the same Spack instance like:

```
spack install openblas
git checkout v0.19
spack install py-numpy ^/openblas-hash #older numpy/python uses newer openblas
git checkout v0.20
```

(right now, we cache repo details that generally require running `spack clean -m` when you use different Spack versions)

**Reply by alalazo** [reply]
> I want to consider whether there is an easy path for users that need to install a mixture of packages needing Python 2.7 and newer packages that may only be available in a later Spack version (i.e. as new packages are added to Spack > 0.19).

Another possible idea, if we have requests of that kind, would be to maintain a separate repository for Python 2.7 packages under the Spack organization. For an approach like this to be feasible we need at least:
1. (Technical request) To move `build-system` definitions from Spack core to repositories (this is because it's very likely that `PythonPackage` will evolve to support Python 3 packages only)
2. (Community request) A few people, other than @adamjstewart, that could maintain the Python 2.7 repository 

Not sure this is the best approach, but if we have some strong case that comes up after deprecation this can be a way to mitigate the impact.

**Comment by scheibelp** [comment]
There was some discussion on this at today's meeting: https://github.com/spack/spack/wiki/Telcon%3A-2022-09-14. The main question I want to bring up here:

Is the plan to also remove any packages that only build with Python 2.7? I think it would make sense to contact the maintainers of those packages (not as a prerequisite for making this change, but as part of it)

**Reply by adamjstewart** [reply]
Yes, that's the plan. Since Python 2 is deprecated, they won't be able to install those packages without noticing that it's deprecated. When I deprecated Python 2 last time, I pinged all of the maintainers of those packages and everyone seemed fine with it.

**Comment by adamjstewart** [comment]
Circling back on this 1 year later. The policy that was adopted after this discussion was:

Regarding (1): Follow [distro support cycle](https://access.redhat.com/product-life-cycles/) (currently Python 3.6+)
Regarding (2): Follow [PSF support cycle](https://endoflife.date/python) (currently Python 3.8+)

However, the fact that (1) and (2) aren't in sync makes it difficult to support bootstrapping Spack (both clingo binaries and style dependencies). I would like to reopen this discussion to find a solution to this issue.

### Proposal

I propose following the PSF support cycle (currently Python 3.8+) for all of Spack. Benefits of this:

1. Developers don't have to use ancient Python syntax. For example, type hints have become much nicer in newer versions of Python.
2. Easy excuse to deprecate and remove older versions of software that don't support any version of Python released in the last 5 years (this is what I've been primarily using this for).
3. Less packages and versions to maintain as a result (I'm almost single-handedly maintaining 2.5K packages in Spack, many with dozens of versions).
4. Fewer tests in GitHub Actions -> faster development, less climate impact.
5. Authoritative source that no one can disagree with (they maintain Python itself).
6. Same deprecation cycle as most other Python libraries, although [NEP-29](https://numpy.org/neps/nep-0029-deprecation_policy.html) is dominant in scientific Python libraries

For users on systems where Python 3.8+ is not installed, they can choose one of the following solutions:

1. Use an older version of Spack
2. Install Python 3.8+
3. Install Python 3.8+ using an older version of Spack

If we release Spack on PyPI, it becomes easier for package managers like pip to only install versions of Spack that are compatible with the Python version being used.

### Alternatives

1. If anyone else wants to take over the above 2.5K packages I'm maintaining, I'm happy to share 🙂 
2. Support 3.6+ for running Spack, but only 3.8+ for bootstrapping its dependencies. Users will have to run `pip install --user clingo` all by themselves 😭 
3. Following the [door-in-the-face technique](https://en.wikipedia.org/wiki/Door-in-the-face_technique), we could adopt [NEP-29](https://numpy.org/neps/nep-0029-deprecation_policy.html) and support Python 3.9+ instead 😉 

Feel free to suggest other alternatives!

@haampie @alalazo @tgamblin

**Reply by haampie** [reply]
> Less packages and versions to maintain as a result (I'm almost single-handedly maintaining 2.5K packages in Spack, many with dozens of versions).

What I fail to see is how removing support for (installing) older Python 3.x is reducing the maintenance burden for Python package maintainers.

You'll get exactly as many review requests as before, and I just cannot imagine more than 1% of the pull requests have to do with support for Python 3.7. You never have to modify rules like

```python
depends_on("python@3.7:", when="@x:")
```

they're frozen.

> Fewer tests in GitHub Actions -> faster development

GHA is certainly not the bottleneck.

---

What I'd really wanna avoid is Spack turning into Nix where you almost need a meta-package manager to figure out what version of Spack has what version of particular software. Spack really has this advantage over Nix, so let's be happy with that.

I also really dread the idea of `pip install <spack / spack dependency>`. We should aim for Spack bootstrapping itself, since it is a much lower level package manager than pip.

I would really want to be able to say to people: with spack you can just login to your 6 year old HPC cluster and run `spack install x` and "it just works".

The suggestion of being on top of deprecating critical packages as soon as possible is not really what I imagine is good for Spack.

**Reply by haampie** [reply]
To take one example that really shows why I am not in favor of this "let's deprecate python 3.7 everywhere" sentiment (from the relevant PR):

![image](https://github.com/spack/spack/assets/194764/c95dcf6b-c6fa-4513-adaf-94d8cec8d15f)

I see absolutely no benefit in these types of changes. No "maintenance burden" type of argument applies here. It's just "let me remove any mention of python 3.7 for consistency". It makes no sense...

**Reply by haampie** [reply]
A more constructive alternative:

Identify packages you find a pain to maintain for Python 3.7 (e.g. branching on 3.7 for specific build instructions, flags, giant patches, ...) and replace the ugliness locally with `depends_on("python@3.8:")`.

**Reply by adamjstewart** [reply]
> What I fail to see is how removing support for (installing) older Python 3.x is reducing the maintenance burden for Python package maintainers.

*Python package _maintainer_

It's not frozen rules I'm worried about, it's ensuring that every new package that gets added still supports Python versions released 7+ years ago.

If your concern is solely about bootstrapping clingo, why don't we add a GHA test that ensures that Spack's dependencies can be bootstrapped? @alalazo tried adding one but it seems to only fail on develop, not on PRs that break bootstrapping.
